### ISSUE TITLE
DatabaseDisconnectError with pgbouncer

### ISSUE THREAD

**Issue by fertobar** [jeremyevans/sequel#1535]
Hello, I have an issue in sequel 5.11.0 with Puma, combined with ELB and pgbouncer.

ELB has a max idle timeout and close the connection after N seconds (default 100 s, max 4000 s)
Then, when there is a peak of traffic new threads are raised and I can see:
`Sequel::DatabaseDisconnectError: PG::ConnectionBad: PQconsumeInput() server closed the connection unexpectedly This probably means the server terminated abnormally before or while processing the request. 
`
that is raised because ELB close connection due idle timeout.

I have **connection_validation_timeout** but seems was ignored.

The unique trick that helped was **connection_expiration_timeout** less than ELB idle timeout. The unique issue with this approach is that refresh of connections is expensive if the period of time is small, with high traffic.

Maybe is an issue related with https://github.com/jeremyevans/sequel/issues/691 , but I'm wornder if there is some way to send some kind of TCP keep alive each n seconds or  different way to configure sequel settings to prevent  the _Sequel::DatabaseDisconnectError_ 
without need to hack Puma or PG driver.

Any help is appreciated.

**Comment by fertobar** [comment]
thanks for the feedback @jeremyevans !

Yes is only an issue with pgbouncer (no present without pgbouncer and DB direct connection) , that I can not bypass since is part of the requirement.

Is also more frequent when there is less activity and traffic start to grow (due more idle connections)

Seems not simple to reproduce in self contained example because only happens when the elb close (reproducible if TLL in ELB is smaller) and multi thread, but I will try to reproduce.

**Comment by jeremyevans** [comment]
If it only happens when using pgbouncer and not with a direct connection, then it is probably a bug in pgbouncer that should be fixed.  Trying to come up with a reproducible example in Sequel is a waste of time, since you already know the problem is elsewhere.  My advice would be communicating with the people in charge to remove the use of pgbouncer as a requirement, since it's causing problems and not providing any advantages.

**Comment by fertobar** [comment]
yes, according I saw seems when ELB close the connection, and at low level the Puma thread think that is still a valid DB connection, but  behind the scene connection was closed.
Anyway I think I can rescue and handle in the aplication level, since is specific scenario.

Thanks for the help!

**Comment by cbandy** [comment]
@fertobar take a look at your TCP keepalive settings in PgBouncer&sup1; and libpq&sup2;. The latter can be set through the `:driver_options` key in `Sequel.connect`.

&sup1; https://pgbouncer.github.io/config.html#tcp_keepalive
&sup2; https://www.postgresql.org/docs/current/static/libpq-connect.html#LIBPQ-KEEPALIVES
&sup3; https://sequel.jeremyevans.net/rdoc/files/doc/opening_databases_rdoc.html#label-postgres

**Comment by fertobar** [comment]
thanks for the feedback @cbandy ! I will try
